### PR TITLE
Feature/ATG-68 Service for game state 

### PR DESCRIPTION
**Pull request by IraCherginyshka**
Create game-state service for whole project
Management information about state of game
Pause/Reset buttons
Change folders for controls

### CODE REVIEW

File: src/app/components/game-board/game-board.component.ts [hunk 1]
```diff
@@ -1,57 +1,81 @@
...
+  private lineWithFigure: number;
+
+  constructor(private gameStateService: GameStateService) {}
 
   ngOnInit(): void {
     this.canvas.nativeElement.width = CANVAS_WIDTH;
     this.canvas.nativeElement.height = CANVAS_HEIGHT;
     this.ctx = this.canvas.nativeElement.getContext('2d');
-    this.boardMatrix = GameBoardComponent.makeBoardEmptyMatrix(
+    this.boardMatrix = MatrixModel.makeBoardEmptyMatrix(
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: Not sure that Matrix model should contain logic for empty matrix generation, i think it was more appropriate place before

File: src/app/components/game-state-controls/game-state-controls.component.html [hunk 2]
```diff
@@ -1,8 +1,8 @@
-<div class="state-controls">
-  <button class="state-control rotation">rotation</button>
-  <div class="horizontal">
-    <button class="state-control left">left</button>
-    <button class="state-control right">right</button>
-  </div>
-  <button class="state-control down">down</button>
+<div class="controls">
+  <button class="controls-action reset" (click)="resetGame()">
```
Comment (oleksandrstarov) [inline src/app/components/game-state-controls/game-state-controls.component.html:2]: these functions are private thus are not accessible from template , it wont build in production mode

File: src/app/components/game-state-controls/game-state-controls.component.ts [hunk 3]
```diff
@@ -1,8 +1,25 @@
...
   templateUrl: './game-state-controls.component.html',
   styleUrls: ['./game-state-controls.component.scss'],
 })
-export class GameStateControlsComponent {}
+export class GameStateControlsComponent {
+  constructor(private gameStateService: GameStateService) {}
+
+  private playGame(): void {
+    this.gameStateService.setGameState(GameState.START);
+  }
```
Comment (oleksandrstarov) [inline src/app/components/game-state-controls/game-state-controls.component.ts]: as a rule we are not creating unused functions

File: src/app/components/game-state-controls/game-state-controls.component.ts [hunk 4]
```diff
@@ -1,8 +1,25 @@
 import { Component } from '@angular/core';
+import { GameState } from '../../enums/game-state.enum';
+
+import { GameStateService } from '../../services/game-state.service';
 
 @Component({
-  selector: 'atg-game-state-controls',
+  selector: 'atg-game-controls',
```
Comment (oleksandrstarov) [inline src/app/components/game-state-controls/game-state-controls.component.ts]: Some mess with selectors here

File: src/app/components/game-controls/game-controls.component.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import { Component } from '@angular/core';
 
 @Component({
-  selector: 'atg-game-controls',
+  selector: 'atg-game-state-controls',
```
Comment (oleksandrstarov) [inline src/app/components/game-controls/game-controls.component.ts]: Please align selector with class name

File: src/app/components/game-board/game-board.component.ts [hunk 6]
```diff
@@ -1,57 +1,81 @@
...
   private play(): void {
-    let height = 0;
+    this.lineWithFigure = 0;
     const newFigure = new FigureModel();
     const itemHeight = newFigure.figureMatrix.length;
-    const interval = setInterval(() => {
-      this.drawBoard(newFigure.showFigure(height, this.boardMatrix));
-      if (height + itemHeight === QUANTITY_BLOCKS_HEIGHT) {
-        clearInterval(interval);
+    const newMatrix = new MatrixModel(this.ctx);
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: I would not create MatrixModel, as it actually not a model but just separate object with one dependency and static function. For me it more obvious to have draw function in this class or in a service

File: src/app/components/game-state-controls/game-state-controls.component.html [hunk 7]
```diff
@@ -1,8 +1,8 @@
...
+    reset
+  </button>
+  <button class="controls-action pause" (click)="pauseGame()">
```
Comment (oleksandrstarov) [inline src/app/components/game-state-controls/game-state-controls.component.html:5]: Same as above

File: src/app/components/game-state-controls/game-state-controls.component.spec.ts [hunk 8]
```diff
@@ -2,7 +2,7 @@ import { async, ComponentFixture, TestBed } from '@angular/core/testing';
 
 import { GameStateControlsComponent } from './game-state-controls.component';
 
-describe('GameStateControlsComponent', () => {
+describe('GameControlsComponent', () => {
```
Comment (oleksandrstarov) [inline src/app/components/game-state-controls/game-state-controls.component.spec.ts]: We have one component as import but another one as test name? Looks like a mistake, please check this

**Review by oleksandrstarov — CHANGES_REQUESTED** [review]

File: src/app/components/game-board/game-board.component.ts (hunk 1 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Ok

File: src/app/components/game-state-controls/game-state-controls.component.html (hunk 2 above)
Comment (IraCherginyshka) [reply src/app/components/game-state-controls/game-state-controls.component.html:2]: Changed

File: src/app/components/game-state-controls/game-state-controls.component.ts (hunk 3 above)
Comment (IraCherginyshka) [reply src/app/components/game-state-controls/game-state-controls.component.ts]: Deleted

File: src/app/components/game-state-controls/game-state-controls.component.ts (hunk 4 above)
Comment (IraCherginyshka) [reply src/app/components/game-state-controls/game-state-controls.component.ts]: Fixed

File: src/app/components/game-controls/game-controls.component.ts (hunk 5 above)
Comment (IraCherginyshka) [reply src/app/components/game-controls/game-controls.component.ts]: Fixed

File: src/app/components/game-board/game-board.component.ts (hunk 6 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Ok

File: src/app/components/game-state-controls/game-state-controls.component.html (hunk 7 above)
Comment (IraCherginyshka) [reply src/app/components/game-state-controls/game-state-controls.component.html:5]: Done

File: src/app/components/game-state-controls/game-state-controls.component.spec.ts (hunk 8 above)
Comment (IraCherginyshka) [reply src/app/components/game-state-controls/game-state-controls.component.spec.ts]: Fixed

File: src/app/components/game-board/game-board.component.ts [hunk 9]
```diff
@@ -29,29 +37,53 @@ export class GameBoardComponent implements OnInit {
...
-  private static makeBoardEmptyMatrix(width: number, height: number): FiguresColors[][] {
+  ngOnDestroy(): void {
+    this.subscriptionState.unsubscribe();
+  }
+  static makeBoardEmptyMatrix(width: number, height: number): FiguresColors[][] {
     return new Array(height).fill(new Array(width).fill(FiguresColors.DEFAULT));
   }
 
+  private drawBoard(matrix: FiguresColors[][]): void {
+    return matrix.forEach((line, indexY) => {
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: so this is void function or it returns FiguresColors[][] ?

File: src/app/components/game-board/game-board.component.ts [hunk 10]
```diff
@@ -29,29 +37,53 @@ export class GameBoardComponent implements OnInit {
...
+        if (gameState === GameState.PAUSE) {
+          this.pause();
+        }
+      });
   }
 
-  private static makeBoardEmptyMatrix(width: number, height: number): FiguresColors[][] {
+  ngOnDestroy(): void {
+    this.subscriptionState.unsubscribe();
+  }
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts:54]: Please separate functions by empty line

**Review by oleksandrstarov — CHANGES_REQUESTED** [review]

File: src/app/components/game-board/game-board.component.ts (hunk 10 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts:54]: Ok

File: src/app/components/game-board/game-board.component.ts (hunk 9 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Fixed

**Review by oleksandrstarov — APPROVED** [review]

**Pull request merged**